### PR TITLE
[Robot] crlp testcase for last soft credit opportunity

### PR DESCRIPTION
**Pull request by rjanjanam**
-Created a page object for object manager and added changes to create a custom field 
-Creating a crlp setting object and rollup for automated soft credit
# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# New Metadata

# Deleted Metadata

### CODE REVIEW

File: robot/Cumulus/resources/ObjectMangerPageObject.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+		url_template = "{root}/lightning/setup/ObjectManager/{object}/FieldsAndRelationships/view"
+		url = url_template.format(root=self.cumulusci.org.lightning_base_url, object=object_name)
+		self.selenium.go_to(url)
+		search_button = npsp_lex_locators['object_manager']['global_search']
+		self.salesforce.wait_until_loading_is_complete()
+		self.selenium.wait_until_page_contains_element(search_button)
+	
+	@capture_screenshot_on_error
+	def create_custom_field(self, type, field_name, related_to):
+		search_button = npsp_lex_locators['object_manager']['global_search']
```
Comment (skristem) [inline robot/Cumulus/resources/ObjectMangerPageObject.py:26]: can u add documentation to this keyword

File: robot/Cumulus/resources/OpportunityPageObject.py [hunk 2]
```diff
@@ -31,6 +32,19 @@ def navigate_to_writeoff_payments_page(self):
         self.npsp.wait_for_locator('frame','Write Off Remaining Balance')
         self.npsp.choose_frame("Write Off Remaining Balance")
         self.selenium.wait_until_page_contains("You are preparing to write off")
+        
+    def change_related_contact_role_settings(self,name,role=None,**kwargs):
+
```
Comment (skristem) [inline robot/Cumulus/resources/OpportunityPageObject.py:37]: add documentation

File: robot/Cumulus/resources/ObjectMangerPageObject.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+				self.selenium.click_element(lookup_locator)
+				time.sleep(1)
+				self.selenium.click_element(next_button)
+				related = npsp_lex_locators['object_manager']['select_related']
+				self.selenium.wait_until_page_contains_element(related,60)
+				self.selenium.scroll_element_into_view(related)
+				self.selenium.get_webelement(related).click()
+				self.selenium.click_element(option)
+				time.sleep(2)
+				self.selenium.click_element(next_button)
```
Comment (skristem) [inline robot/Cumulus/resources/ObjectMangerPageObject.py:59]: I am guessing the reason for sleep statements is there is no change to elements that can be waited upon for verification

File: robot/Cumulus/resources/OpportunityPageObject.py [hunk 4]
```diff
@@ -31,6 +32,19 @@ def navigate_to_writeoff_payments_page(self):
...
+        
+    def change_related_contact_role_settings(self,name,role=None,**kwargs):
+        
+        dropdown = npsp_lex_locators['related_drop_down'].format(name)
+        edit = npsp_lex_locators['record']['dd_edit_option'].format("Edit")
+        self.selenium.wait_until_page_contains_element(dropdown)
+        self.salesforce._jsclick(dropdown)
+        self.selenium.wait_until_element_is_visible(edit)
+        self.selenium.click_element(edit)
+        time.sleep(2)
```
Comment (skristem) [inline robot/Cumulus/resources/OpportunityPageObject.py:44]: can we check on something instead of adding sleep?

File: robot/Cumulus/resources/locators_48.py [hunk 5]
```diff
@@ -6,6 +6,7 @@
     'breadcrumb-link':"//a[@title='{}' and contains(@class,'BreadCrumbItem')]",
     'obj-header':"//h1//*[text()='{}']",
     'button-title':"//button[@title='{}']",
+    'button-with-text':'//button[contains(text(),"{}")]',
```
Comment (skristem) [inline robot/Cumulus/resources/locators_48.py:9]: curious, we already have a locator for //button[text()='{}']locator, can't we just use the same

File: robot/Cumulus/resources/locators_48.py [hunk 6]
```diff
@@ -232,6 +235,12 @@
 
     'object_manager':{
         'button':'//input[@title="{}"]',
+        'Lookup_option':'//td/input[@id = "dtypeY"]',
+        'select_related':'//select[@id = "DomainEnumOrId"]',
+        'select_related_option':'//select[@id = "DomainEnumOrId"]/option[@value="{}"]',
+        'input_field_label': '//input[@id="MasterLabel"]',
+        'global_search':'//input[@id="globalQuickfind"]',
```
Comment (skristem) [inline robot/Cumulus/resources/locators_48.py:242]: can line 238, 241 and 242 use same locator //input[@id='{}'] and pass id in the keyword def

File: robot/Cumulus/resources/locators_48.py [hunk 7]
```diff
@@ -254,6 +263,12 @@
         'new_button': '//button[text() = "{}"]',
         'select_locator': "//select[@name ='{}']",
         'success_toast': "//div[contains(@class,'slds-notify.slds-notify_toast,slds-theme_success')]",
+        'target_option': "//select[@name = 'Target Field']/option[text()='{}']",
+        'target': "//select[@name = 'Target Field']",
+        'target_object': "//select[@name = 'Target Object']",
```
Comment (skristem) [inline robot/Cumulus/resources/locators_48.py:268]: these could be changed to 1 by passing arguments in keyword, same for 266 and 269

File: robot/Cumulus/tests/browser/crlps/last_soft_credit_opportunity.robot [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+***Keywords***
+# creates test data a contact and an opportunity for the contact
+Setup Test Data
+    Setupdata   contact   ${contact1_fields}     ${opportunity_fields}
+    ${date} =   Get Current Date       result_format=%-m/%-d/%Y
+    Set suite variable    ${date}
+
+# Navigates to Object Manger for the specified object and checks for the presence of the custom field
+# If the specified custom field is not present adds the custom field
```
Comment (skristem) [inline robot/Cumulus/tests/browser/crlps/last_soft_credit_opportunity.robot:26]: instead of comments like this, u could add this as documentation for the keyword. Also could u move this to .robot file with arguments so that I could use same keyword for my test aswell

File: robot/Cumulus/tests/browser/crlps/last_soft_credit_opportunity.robot [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+
+*** Variables ***
+&{contact1_fields}       Email=test@example.com
+&{opportunity_fields}    Type=Donation   Name=Customizable rollup test $100 Donation   Amount=100  StageName=Closed Won  npe01__Do_Not_Automatically_Create_Payment__c=false
+
+*** Test Cases ***
+
+Create Crlp For Automated Soft Credit
+    [Documentation]
```
Comment (skristem) [inline robot/Cumulus/tests/browser/crlps/last_soft_credit_opportunity.robot:42]: plz add documentation for test case

File: robot/Cumulus/resources/ContactPageObject.py [hunk 10]
```diff
@@ -64,8 +64,12 @@ def waitfor_actions_dropdown_and_click_option(self,option):
            Click on the desired option passed as a parameter
         """
         loc=npsp_lex_locators['contacts_actions_dropdown_menu']
+        self.npsp.click_more_actions_button()
         self.selenium.wait_until_element_is_visible(loc)
         self.selenium.click_link(option)
+        if option == 'Recalculating Rollups':
+            notification = npsp_lex_locators['crlps']['rollup_progress_notification'].format("rollups batch job is calculating")
+            self.selenium.wait_until_element_is_visible(notification)
```
Comment (skristem) [inline robot/Cumulus/resources/ContactPageObject.py:72]: I dont see this being used anywhere, are these changes required?

Comment (rjanjanam) [reply robot/Cumulus/resources/ContactPageObject.py]: Removed it. Thought we would need it earlier but not using

File: robot/Cumulus/resources/ObjectMangerPageObject.py (hunk 1 above)
Comment (rjanjanam) [reply robot/Cumulus/resources/ObjectMangerPageObject.py:55]: Done

File: robot/Cumulus/resources/locators_48.py (hunk 5 above)
Comment (rjanjanam) [reply robot/Cumulus/resources/locators_48.py:9]: This requires contains because the text is not exact and has spaces in between. Trim did not work either.

File: robot/Cumulus/resources/locators_48.py (hunk 6 above)
Comment (rjanjanam) [reply robot/Cumulus/resources/locators_48.py:242]: This gets slightly complicated and not so readable because each of them comes with a unique id like dtypeY, dtypeX, so from the code you will have to format option.format(dtypeX) etc.

File: robot/Cumulus/resources/locators_48.py (hunk 7 above)
Comment (rjanjanam) [reply robot/Cumulus/resources/locators_48.py]: Actually realized I am not using them so deleted them

File: robot/Cumulus/resources/locators_48.py (hunk 6 above)
Comment (skristem) [reply robot/Cumulus/resources/locators_48.py:242]: yes by doing that u r making ur locators unique for that keyword. If u hardcode ids in the locator's file, we will not be making our locators reusable but we will end up with lots of locators in our locator file. Levels dictionary already contains a locator for input with id, please see if you can repurpose it and if yes move it to common locators and make adjustments for existing as well as new test

File: robot/Cumulus/tests/browser/crlps/last_soft_credit_opportunity.robot (hunk 9 above)
Comment (rjanjanam) [reply robot/Cumulus/tests/browser/crlps/last_soft_credit_opportunity.robot]: done

File: robot/Cumulus/resources/locators_48.py (hunk 6 above)
Comment (rjanjanam) [reply robot/Cumulus/resources/locators_48.py]: Updated to a single locator

File: robot/Cumulus/resources/ObjectMangerPageObject.py (hunk 3 above)
Comment (rjanjanam) [reply robot/Cumulus/resources/ObjectMangerPageObject.py]: yes

File: robot/Cumulus/resources/NPSP.robot [hunk 11]
```diff
@@ -331,10 +316,24 @@ Enable Customizable Rollups
...
-    [Documentation]    This keyword enables advanced mapping if not already enabled.  
+    [Documentation]    This keyword enables advanced mapping if not already enabled.
     Go To Page                              Custom          NPSP_Settings
     Open Main Menu                          System Tools
-    Click Link With Text                    Advanced Mapping for Data Import & Gift Entry   
-    Enable Advanced Mapping If Not Enabled 
+    Click Link With Text                    Advanced Mapping for Data Import & Gift Entry
+    Enable Advanced Mapping If Not Enabled
+
+    Enable Customizable Rollups If Not Enabled
```
Comment (skristem) [inline robot/Cumulus/resources/NPSP.robot:327]: please remove this line from Enable Advanced Mapping keyword. Looks like u missed it while resolving conflicts

File: robot/Cumulus/resources/OpportunityPageObject.py (hunk 4 above)
Comment (rjanjanam) [reply robot/Cumulus/resources/OpportunityPageObject.py]: Agree and changed to wait_for_modal

File: robot/Cumulus/resources/locators_48.py [hunk 12]
```diff
@@ -232,6 +235,10 @@
 
     'object_manager':{
         'button':'//input[@title="{}"]',
+		'input':'//input[@id="{}"]',
```
Comment (skristem) [inline robot/Cumulus/resources/locators_48.py:238]: can u fix spacing here
